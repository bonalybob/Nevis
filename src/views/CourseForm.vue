<template>
  <main>
    <h1 class="mx-10 mb-1">
      <template v-if="$route.path.includes('create')">
        <back-arrow :to="`/events/${$route.params.eventId}/courses`" />Create Course
      </template>
      <template v-else>
        <back-arrow />Update Course
      </template>
    </h1>
    <div v-if="$route.path.includes('create')" class="mx-12 mb-3">
      <button class="button" @click="submit">Create Course</button>
    </div>
    <div v-else class="mx-12 mb-3">
      <button class="button" @click="submit">Update Course</button>
      <button class="button" @click="showConfirmationDialog = true">Delete Course</button>
    </div>
    <form class="shadow mx-12" @submit.prevent="submit">
      <text-input v-model.trim="course.name" label="Name:" />
      <text-input v-model="course.length" label="Length (km): " />
      <text-input v-model="course.climb" label="Climb (m):" />
      <text-input v-model.trim="course.controls" label="Controls:" />
    </form>
    <transition name="fade">
      <confirmation-dialog
        v-if="showConfirmationDialog"
        heading="Delete Course"
        message="Are You Sure You Want to Delete This Course? This Action Can't Be Recovered."
        confirm="Delete"
        cancel="Cancel"
        @close="onConfirm"
      />
    </transition>
  </main>
</template>

<script>
import BackArrow from '@/components/BackArrow'
import TextInput from '@/components/TextInput'
import ConfirmationDialog from '@/components/ConfirmationDialog'

export default {
  components: {
    'back-arrow': BackArrow,
    'text-input': TextInput,
    'confirmation-dialog': ConfirmationDialog,
  },

  data: function () {
    return {
      showConfirmationDialog: false,
      course: {
        name: '',
        length: 0,
        climb: 0,
        controls: '',
        id: undefined,
      },
    }
  },

  mounted: function () {
    if (this.$database.connection === null || !this.$database.connected) {
      this.$router.push('/')
      this.$messages.addMessage('Problem Connecting To Database', 'error')
    }
    else if (this.$route.params && this.$route.params.courseId) this.getCourseDetails()
  },

  methods: {
    submit: async function () {
      if (await this.checkForDuplicateCourse()) {
        this.$messages.addMessage(`Course "${this.course.name}" Already Exists`, 'error')
      }
      else if (!this.course.controls.match(/^[0-9,\s]*$/) || this.course.controls.match(/^.*,$/)) {
        this.$messages.addMessage(`Invalid Format for Controls`, 'error')
        this.$messages.addMessage(`Please Separate Each Code With a Comma e.g. 101, 102, 103`)
      }
      else if (typeof this.course.length === 'string' && !this.course.length.match(/^[0-9]*\.*[0-9]*$/)) {
        this.$messages.addMessage(`Please enter a number for Course Length`, 'error')
      }
      else if (typeof this.course.climb === 'string' && !this.course.climb.match(/^[0-9]*\.*[0-9]*$/)) {
        this.$messages.addMessage(`Please enter a number for Course Climb`, 'error')
      }
      else if (this.course.id) this.updateCourse()
      else this.createCourse()
    },

    getCourseDetails: function () {
      return this.$database.query('SELECT * FROM courses WHERE id=? LIMIT 1', this.$route.params.courseId)
        .then(result => {
          if (result && result[0]) {
            this.course = result[0]
            if (result[0].length) this.course.length = this.course.length / 1000
          }
          else {
            this.$messages.addMessage('Problem Fetching Course Data', 'error')
          }
        })
        .catch(() => this.$messages.addMessage('Problem Fetching Course Data', 'error'))
    },

    createCourse: function () {
      return this.$database.query('INSERT INTO courses SET ?', {
        name: this.course.name,
        event: this.$route.params.eventId,
        length: this.course.length * 1000 || 0,
        climb: this.course.climb || 0,
        controls: this.course.controls,
        type: 'linear',
      })
        .then(() => this.$router.push(`/events/${this.$route.params.eventId}/courses`))
        .catch(() => this.$messages.addMessage('Problem Creating Course', 'error'))
    },

    updateCourse: function () {
      return this.$database.query('UPDATE courses SET ? WHERE id=?', [{
        id: this.course.id,
        name: this.course.name,
        event: this.$route.params.eventId,
        length: this.course.length * 1000 || 0,
        climb: this.course.climb || 0,
        controls: this.course.controls,
        type: 'linear',
      }, this.course.id])
        .then(() => this.$router.push(`/events/${this.$route.params.eventId}/courses`))
        .catch(() => this.$messages.addMessage('Problem Updating Course', 'error'))
    },

    deleteCourse: function () {
      return this.$database.query('DELETE FROM courses WHERE id=?', this.course.id)
        .then(() => {
          this.$messages.addMessage(`Course "${this.course.name}" Deleted`)
          this.$router.push(`/events/${this.$route.params.eventId}/courses`)
        })
        .catch(() => this.$messages.addMessage('Problem Deleting Course', 'error'))
    },

    checkForDuplicateCourse: async function () {
      const queryResult = await this.$database.query('SELECT * FROM courses WHERE name=? AND event=?', [this.course.name, parseInt(this.$route.params.eventId)])
      if (this.course.id && queryResult[0] && queryResult[0].id === this.course.id) return false
      else return queryResult.length > 0
    },

    onConfirm: function (decision) {
      this.showConfirmationDialog = false
      if (decision) this.deleteCourse()
    },
  },
}

</script>

