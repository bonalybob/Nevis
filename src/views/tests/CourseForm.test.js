import { mount, shallowMount } from '@vue/test-utils'
import CourseForm from '@/views/CourseForm'

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - Create', () => {
  const wrapper = shallowMount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Update', () => {
  const wrapper = shallowMount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue({ id: 0, name: 'Test', length: 1000, climb: 2 }) },
      $route: { params: { eventId: 12 }, path: '/events/edit/12' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setData({ course: { id: 0, name: 'Test', length: 1000, climb: 2 } })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: false, query: jest.fn().mockResolvedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Get Course Details - Success', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ id: 0, name: 'Test', length: 1000, climb: 2 }]) },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourseDetails()
  expect(wrapper.vm.course).toEqual({ id: 0, name: 'Test', length: 1, climb: 2 })
})

test('Get Course Details - No Length', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ id: 0, name: 'Test', climb: 2 }]) },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourseDetails()
  expect(wrapper.vm.course).toEqual({ id: 0, name: 'Test', climb: 2 })
})

test('Get Course Details - No Data', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourseDetails()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Course Data', 'error')
  expect(wrapper.vm.course.name).toBe('')
})

test('Get Course Details - Error', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourseDetails()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Course Data', 'error')
  expect(wrapper.vm.course.name).toBe('')
})

test('Create Course - Success', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setData({ course: { name: 'Test', date: '' } })
  await wrapper.vm.createCourse()
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12/courses')
})

test('Create Course - Error', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.createCourse()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Creating Course', 'error')
})

test('Update Course - Success', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ id: 0, name: 'Test', length: 1000, climb: 2 }]) },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.updateCourse()
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12/courses')
})

test('Update Course - Error', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.updateCourse()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Updating Course', 'error')
})

test('Delete Course - Success', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue([{ name: 'Test', id: 6 }]) },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setData({ course: { id: 0, name: 'Test', length: 1000, climb: 2 } })
  await wrapper.vm.deleteCourse()
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/events/12/courses')
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(2)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Course "Test" Deleted')
})

test('Delete Course - Error', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.deleteCourse()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Deleting Course', 'error')
})

test('Check for Duplicates', async () => {
  const mockDatabaseQuery = jest.fn().mockResolvedValue([{ name: 'Test', id: 6 }])
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: mockDatabaseQuery },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  mockDatabaseQuery.mockResolvedValue([])
  expect(await wrapper.vm.checkForDuplicateCourse()).toBeFalsy()

  wrapper.setData({ course: { id: null } })
  mockDatabaseQuery.mockResolvedValue([{ id: 1 }, { id: 2 }, { id: 3 }])
  expect(await wrapper.vm.checkForDuplicateCourse()).toBeTruthy()

  wrapper.setData({ course: { id: 12 } })
  mockDatabaseQuery.mockResolvedValue([{ id: 12 }])
  expect(await wrapper.vm.checkForDuplicateCourse()).toBeFalsy()

  mockDatabaseQuery.mockResolvedValue([{ id: 1 }, { id: 2 }, { id: 3 }])
  expect(await wrapper.vm.checkForDuplicateCourse()).toBeTruthy()
})

test('Submit', async () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setMethods({
    checkForDuplicateCourse: jest.fn(),
    updateCourse: jest.fn(),
    createCourse: jest.fn(),
  })

  wrapper.setData({ course: { name: 'Test' } })
  wrapper.vm.checkForDuplicateCourse.mockResolvedValue(true)
  await wrapper.vm.submit()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Course "Test" Already Exists', 'error')

  wrapper.setData({ course: { controls: 'ab, cd' } })
  wrapper.vm.checkForDuplicateCourse.mockResolvedValue(false)
  await wrapper.vm.submit()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(3)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Please Separate Each Code With a Comma e.g. 101, 102, 103')

  wrapper.setData({ course: { controls: '12,23,' } })
  await wrapper.vm.submit()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(5)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Please Separate Each Code With a Comma e.g. 101, 102, 103')

  wrapper.setData({ course: { controls: '12,23', length: 'hello' } })
  await wrapper.vm.submit()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(6)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Please enter a number for Course Length', 'error')

  wrapper.setData({ course: { length: '3.5', climb: 'hello' } })
  await wrapper.vm.submit()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(7)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Please enter a number for Course Climb', 'error')

  wrapper.setData({ course: { id: 1, climb: '2' } })
  await wrapper.vm.submit()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledTimes(7)
  expect(wrapper.vm.createCourse).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.updateCourse).toHaveBeenCalledTimes(1)

  wrapper.setData({ course: { id: undefined } })
  await wrapper.vm.submit()
  expect(wrapper.vm.createCourse).toHaveBeenCalledTimes(1)
})

test('On Confirm', () => {
  const wrapper = mount(CourseForm, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setMethods({ deleteCourse: jest.fn() })
  wrapper.vm.onConfirm(false)
  expect(wrapper.vm.deleteCourse).toHaveBeenCalledTimes(0)
  wrapper.vm.onConfirm(true)
  expect(wrapper.vm.deleteCourse).toHaveBeenCalledTimes(1)
})
