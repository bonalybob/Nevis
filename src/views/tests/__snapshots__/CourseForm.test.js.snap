// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - Create 1`] = `
<main>
  <h1
    class="mx-10 mb-1"
  >
    <back-arrow-stub
      to=""
    />
    Update Course
    
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <button
      class="button"
    >
      Update Course
    </button>
     
    <button
      class="button"
    >
      Delete Course
    </button>
  </div>
   
  <form
    class="shadow mx-12"
  >
    <text-input-stub
      label="Name:"
      type=""
      value=""
    />
     
    <text-input-stub
      label="Length (km): "
      type=""
      value="0"
    />
     
    <text-input-stub
      label="Climb (m):"
      type=""
      value="0"
    />
     
    <text-input-stub
      label="Controls:"
      type=""
      value=""
    />
  </form>
</main>
`;

exports[`Renders Correctly - Update 1`] = `
<main>
  <h1
    class="mx-10 mb-1"
  >
    <back-arrow-stub
      to=""
    />
    Update Course
    
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <button
      class="button"
    >
      Update Course
    </button>
     
    <button
      class="button"
    >
      Delete Course
    </button>
  </div>
   
  <form
    class="shadow mx-12"
  >
    <text-input-stub
      label="Name:"
      type=""
      value="Test"
    />
     
    <text-input-stub
      label="Length (km): "
      type=""
      value="1000"
    />
     
    <text-input-stub
      label="Climb (m):"
      type=""
      value="2"
    />
     
    <text-input-stub
      label="Controls:"
      type=""
      value=""
    />
  </form>
</main>
`;
